blog/2023-11-01-feature-flags-sdks-architectures.md

```diff
@@ -0,0 +1,127 @@
+
+### API endpoints requests with cache using Relay Proxy - example flow
+
+<img src={require('@site/static/img/blog/feature-flags-sdks-architectures/API-endpoints-requests-with-cache-using-Relay-Proxy.png').default} />
+
+#### Advantages
+
+* Network traffic can be reduced.
+
+#### Disadvantages
+
+* Functional disadvantage is that intermediate issue like network or feature flags service error for non-cached data is handled by not evaluating to the expected feature flag value.
+* Another functional disadvantage is that intermediate issue like network or feature flags service error during microservice initialization is handled by not evaluating to the expected feature flag value.
+* First requests are not in cache, thus slower and causing network traffic.
+* Not every request can be efficiently cached, as some evaluations can be dynamic according to context with different values.
+* When there is no background task for getting real-time configuration changes event and refreshing the cache, values can take time to be updated.
+
+### Local evaluation
+
+With this approach, the feature flag configuration is saved locally at the SDK.  
+Feature flag evaluation is done locally via the SDK and not affecting network traffic.
+With this approach, configuration can be fetched on initialization, and be refreshed periodically and/or when there is a configuration change, which can be triggered at the SDK via stream event from the feature flag service. 
+Some SDKs sync data via WebSockets or [SSE](https://en.wikipedia.org/wiki/Server-sent_events).
+When this approach is used, evaluation analytics metrics can be aggregated at the SDK side and published separately, as not every flag evaluation triggers a message to the Features Service / Proxy.
+
+### Local evaluation - example flow
+
+<img src={require('@site/static/img/blog/feature-flags-sdks-architectures/Local-evaluation.png').default} />
+
+#### Advantages
+
+* Functional advantage is that intermediate issue like network or feature flag service errors are automatically handled, as configuration is cached, and flags are evaluated locally.
+  If an issue like network error is ongoing for a long time period, SDK provider state can also be reflected as stale.
+* Another functional advantage, initialization can have a fallback mechanism to dealing with situations like temporary network failure by starting from last received configuration or some other configuration, thus microservice can start with last feature flag configuration.
+* Feature flag configuration can be cached on CDN.
+* Feature flag evaluation is done locally via the SDK and not affecting network traffic, hence can be fast and efficient.
+* Working without a relay proxy can still be efficient from SDK point of view, as flags are evaluated locally and network traffic does not block common flow.
+
+#### Disadvantages
+
+* When feature flag configuration is very large with many feature flags, network traffic and local cache sizes can be increased.
```

Comment:
```suggestion
* When feature flag configurations are very large with many feature flags, network traffic and local cache sizes can be increased. This can impact the time an SDK needs to initialize.
```